### PR TITLE
Refactor our CircleCI workflows

### PR DESCRIPTION
**Pull request by dannycoates**
In order to reduce our CI resource usage we've combined some jobs and changed what runs for each workflow.

Most packages have fast test runs. We've combined these under one job. Packages with slower runs use their own job so they can be run in parallel. Also, we've added branch filters to the rarely changing, but resource intensive jobs.

We now have three workflows that run at different times:
    
1. test_pull_request: runs only on PR branches and only runs tests.
2. deploy_master: runs only on master and only builds and deploys docker images
3. test_and_deploy_tag: runs only on tags and tests, builds, and deploys

 ## Review notes

- `config.yml` is the biggie.
- I made a few small changes to some packages, but only to get tests passing.

### CODE REVIEW

File: .circleci/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# CircleCI
```
Comment (vbudhram) [inline .circleci/README.md]: Thanks for documenting this!

File: packages/fxa-profile-server/test/api.js [hunk 2]
```diff
@@ -875,8 +875,8 @@ describe('/avatar', function() {
 
   describe('upload', function() {
     it('should upload a new avatar', function() {
-      this.slow(2000);
-      this.timeout(3000);
+      this.slow(5000);
```
Comment (vbudhram) [inline packages/fxa-profile-server/test/api.js]: Did the smaller instance size cause this to be slower?

File: packages/fxa-event-broker/package.json [hunk 3]
```diff
@@ -8,10 +8,10 @@
     "audit": "npm audit --json | audit-filter --nsp-config=.nsprc --audit=-",
     "lint:tslint": "./node_modules/tslint/bin/tslint -p .",
     "watch": "tsc -w",
-    "test": "FIRESTORE_EMULATOR_HOST=localhost:8006 FIRESTORE_PROJECT_ID=fx-event-broker ./node_modules/mocha/bin/mocha -r ts-node/register test/**/*.spec.ts test/**/**/*.spec.ts",
+    "test": "mocha -r ts-node/register test/**/*.spec.ts test/**/**/*.spec.ts",
     "start-prod": "npm run build && node ./dist/bin/worker.js",
-    "start": "npm run build && NODE_ENV=development PUBSUB_EMULATOR_HOST=localhost:8005 FIRESTORE_EMULATOR_HOST=localhost:8006 node ./dist/bin/workerDev.js",
-    "start-dev-debug": "NODE_ENV=development PUBSUB_EMULATOR_HOST=localhost:8005 FIRESTORE_EMULATOR_HOST=localhost:8006 nodemon --legacy-watch --watch ./dist --inspect=0.0.0.0:9180 --nolazy ./dist/bin/workerDev.js"
+    "start": "npm run build && NODE_ENV=development node ./dist/bin/workerDev.js",
+    "start-dev-debug": "NODE_ENV=development nodemon --legacy-watch --watch ./dist --inspect=0.0.0.0:9180 --nolazy ./dist/bin/workerDev.js"
```
Comment (bbangert) [inline packages/fxa-event-broker/package.json]: I was under the impression those env vars need to be set or it won't use the emulators. That isn't the case after all?

File: packages/fxa-profile-server/test/api.js (hunk 2 above)
Comment (dannycoates) [reply packages/fxa-profile-server/test/api.js]: yes. it was right on the edge and failed sometimes on a `small`

File: packages/fxa-event-broker/package.json (hunk 3 above)
Comment (dannycoates) [reply packages/fxa-event-broker/package.json]: It looks like the ports are hard coded in the tests. I didn't want to dig too deep.

**Comment by dannycoates** [comment]
Thanks, y'all. I'm going to rebase and try enabling custom-server and event-broker

**Comment by lmorchard** [comment]
FWIW, whenever this is done & merged, remember to update the ecosystem docs 😅 

https://github.com/mozilla/ecosystem-platform/blob/master/docs/fxa-engineering/fxa-tests-circleci.md

File: .circleci/README.md (hunk 1 above)
Comment (lmorchard) [reply .circleci/README.md]: Actually, could this mostly replace the ecosystem site doc? Lots of per-job detail that that seems to go away now

Comment (dannycoates) [reply .circleci/README.md]: Thanks, Les! I'll update that next. Maybe just to point here

File: packages/fxa-event-broker/package.json (hunk 3 above)
Comment (dannycoates) [reply packages/fxa-event-broker/package.json]: i put the env vars back

**Pull request merged**